### PR TITLE
feat: Add Support for Tasklist

### LINKED ISSUE #40: feat: Add support for TaskList
_Need to add feature of Adding, creating, deleting and display tasks added by user as implemented in [mern-boilerplate](https://boilerplate-mern.platform.jalantechnologies.com/tasks)._

Can make use of SwipeList with NativeBase as described [here](https://docs.nativebase.io/building-swipe-list) to make the tasks more interactive and user engaging.

### PR DESCRIPTION
**Pull request by abhishek310001**
## Description
_The goal of this PR is add the feature of displaying tasklist on the Dashboard screen of the app._

## Screenshots
| Android | iOS |
| ---- | ---- |
| Screenshot_1723531397 | Simulator Screenshot - iPhone 15 Pro Max - 2024-08-13 at 12 11 18 |
| Screenshot_1723531401 | Simulator Screenshot - iPhone 15 Pro Max - 2024-08-13 at 12 11 22 |
| Screenshot_1723531423 | Simulator Screenshot - iPhone 15 Pro Max - 2024-08-13 at 12 11 37 |
| Screenshot_1723531428 | Simulator Screenshot - iPhone 15 Pro Max - 2024-08-13 at 12 11 43 |
| Screenshot_1723531450 | Simulator Screenshot - iPhone 15 Pro Max - 2024-08-13 at 12 12 03 |
| Screenshot_1723531456 | Simulator Screenshot - iPhone 15 Pro Max - 2024-08-13 at 12 40 00 |
| Screenshot_1723533266 | Simulator Screenshot - iPhone 15 Pro Max - 2024-08-13 at 12 41 30 |
| Screenshot_1723533442 | Simulator Screenshot - iPhone 15 Pro Max - 2024-08-13 at 12 42 32 |
| Screenshot_1723533273| Simulator Screenshot - iPhone 15 Pro Max - 2024-08-13 at 12 41 35 |

## Video
_Video demonstrating usage of this feature._

https://github.com/user-attachments/assets/efaeca9d-eca0-453c-b24b-882b1fa7c313

_Video demonstrating various error and validation of the feature._

https://github.com/user-attachments/assets/aaaa668f-59c3-4846-a6e1-e9d673e9d83b

### Automated test cases added
NA

## Devices Tested
- [x] iPhone 15 Pro Max (iOS 17.5))
- [x] Google Pixel 7 (Android 14)

### Manual test cases run
_For each manual test case, list the steps to test or reproduce the PR._
TBA

### CODE REVIEW

**Comment by bbbneo333** [comment]
@abhishek310001 - Will do a through review in sometime. Initial thoughts - the swipe UI does not look right. The entire row should be swipable as one single row, but it looks like upon swiping the content moves down to another row. Anything we can do about that? Once fixed, please do post a video of the exp. for review. Thanks

File: src/constants/task.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  ADD = 'Add',
+  EDIT = 'Edit',
+}
+
+export const TaskValidationLimits = {
+  TITLE_MIN_LENGTH: 3,
+  DESCRIPTION_MIN_LENGTH: 10,
+};
+
+export const TaskValidationMessages = {
```
Comment (bbbneo333) [inline src/constants/task.ts]: Generally - Locale files are used to store user facing messages. Based on the set locale (example - en - english), the respective message is then shown.

Do we have locale support in this app? If yes - We should use that instead. If not - We need add locale support first then.

File: src/services/task.service.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import { AccessToken, APIResponse } from '../types';
+import { Task } from '../types';
+
+import { APIService } from './api-service';
+
+export class TaskService extends APIService {
+  private getAuthHeaders(userAccessToken: AccessToken) {
```
Comment (bbbneo333) [inline src/services/task.service.ts]: Think this can be made part of `APIService`. It can be then reused.

File: src/contexts/task-context.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+export const TaskContextProvider: React.FC<PropsWithChildren> = ({ children }) => {
+  const taskService = useMemo(() => new TaskService(), []);
+
+  const [tasks, setTasks] = useState<Task[]>([]);
+  const [isTasksLoading, setIsTasksLoading] = useState(false);
+  const [isAddTaskLoading, setIsAddTaskLoading] = useState(false);
+  const [isUpdateTaskLoading, setIsUpdateTaskLoading] = useState(false);
+  const [isDeleteTaskLoading, setIsDeleteTaskLoading] = useState(false);
+
+  const { getAccessTokenFromStorage } = useAuthContext();
```
Comment (bbbneo333) [inline src/contexts/task-context.tsx]: Though this is not related to this change, let's rename `getAccessTokenFromStorage` to `getAccessToken` instead. The caller does not needs to know from _where_ (ex - local storage) the token is being fetched from from. Same can be done for related functionalities:

- setAccessTokenToStorage -> setAccessToken
- clearAccessTokenFromStorage -> deleteAccessToken

File: src/contexts/task-context.tsx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    const newTask = new Task({ ...data });
+    setTasks(prevTasks => [newTask, ...prevTasks]);
+    setIsAddTaskLoading(false);
+  };
+
+  const updateTask = async (task: Task) => {
+    setIsUpdateTaskLoading(true);
+    const { error } = await taskService.updateTask(
+      task,
+      getAccessTokenFromStorage() as AccessToken,
```
Comment (bbbneo333) [inline src/contexts/task-context.tsx]: This assertion `as AccessToken` is being repeated everywhere. Can we update the function to define the type so we don't have to do this whenever it's being used?

**Review by bbbneo333 — CHANGES_REQUESTED** [review]
Overall looks good. Added few comments around existing functionality.

File: src/constants/task.ts (hunk 1 above)
Comment (abhishek310001) [reply src/constants/task.ts]: Yeah, we do have locale support in the app. Added these validation messages to the local file.

File: src/services/task.service.ts (hunk 2 above)
Comment (abhishek310001) [reply src/services/task.service.ts]: Implemented.

File: src/contexts/task-context.tsx (hunk 3 above)
Comment (abhishek310001) [reply src/contexts/task-context.tsx]: Renamed them.

File: src/contexts/task-context.tsx (hunk 4 above)
Comment (abhishek310001) [reply src/contexts/task-context.tsx]: Did a workaround and removed this repetition of assertion from everywhere.

File: src/constants/task.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  ADD = 'Add',
+  EDIT = 'Edit',
+}
+
+export const TaskValidationLimits = {
+  TITLE_MIN_LENGTH: 3,
+  DESCRIPTION_MIN_LENGTH: 10,
+};
+
+export const TaskMessages = {
```
Comment (bbbneo333) [inline src/constants/task.ts]: These messages are still here. Should be either hardcoded into UI components or loaded from locale files.

File: src/services/account.service.ts [hunk 6]
```diff
@@ -4,11 +4,10 @@ import { APIService } from './api-service';
...
 export class AccountService extends APIService {
   getAccount = async (userAccessToken: AccessToken): Promise<APIResponse> => {
-    return this.get(`/accounts/${userAccessToken.accountId}`, {
-      headers: {
-        Authorization: `Bearer ${userAccessToken.token}`,
-      },
-    });
+    return this.get(
+      `/accounts/${userAccessToken.accountId}`,
+      this.getBearerTokenHeader(userAccessToken.token),
```
Comment (bbbneo333) [inline src/services/account.service.ts]: Previous name was okay - `getAuthorizationHeader`. `Bearer` is an internal implementation detail that caller does not needs to be aware about.

Comment (abhishek310001) [reply src/services/account.service.ts]: Okay, got it.

File: src/constants/task.ts (hunk 5 above)
Comment (abhishek310001) [reply src/constants/task.ts]: removed these too

**Review by bbbneo333 — APPROVED** [review]

**Pull request merged**